substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -2266,6 +2382,384 @@ impl<T: Config> Pallet<T> {
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        // Only service or consumer can create contract
+        ensure!(
+            caller_twin_id == service_twin_id || caller_twin_id == consumer_twin_id,
+            Error::<T>::TwinNotAuthorizedToCreateServiceContract,
+        );
+
+        // Service twin and consumer twin can not be the same
+        ensure!(
+            service_twin_id != consumer_twin_id,
+            Error::<T>::ServiceContractCreationNotAllowed,
+        );
+
+        // Create service contract
+        let service_contract = types::ServiceContract {
+            service_twin_id,
+            consumer_twin_id,
+            base_fee: 0,
+            variable_fee: 0,
+            metadata: vec![].try_into().unwrap(),
+            accepted_by_service: false,
+            accepted_by_consumer: false,
+            last_bill: 0,
+            state: types::ServiceContractState::Created,
+            phantom: PhantomData,
+        };
+
+        // Get the service contract ID map and increment
+        let mut id = ServiceContractID::<T>::get();
+        id = id + 1;
+
+        // Insert into service contract map
+        ServiceContracts::<T>::insert(id, &service_contract);
+
+        // Update Contract ID
+        ServiceContractID::<T>::put(id);
+
+        // Trigger event for service contract creation
+        Self::deposit_event(Event::ServiceContractCreated {
+            service_contract_id: id,
+        });
+
+        Ok(().into())
+    }
+
+    pub fn _service_contract_set_metadata(
+        account_id: T::AccountId,
+        service_contract_id: u64,
+        metadata: Vec<u8>,
+    ) -> DispatchResultWithPostInfo {
+        let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
+            .ok_or(Error::<T>::TwinNotExists)?;
+
+        let mut service_contract = ServiceContracts::<T>::get(service_contract_id)
+            .ok_or(Error::<T>::ServiceContractNotExists)?;
+
+        // Only service or consumer can set metadata
+        ensure!(
+            twin_id == service_contract.service_twin_id
+                || twin_id == service_contract.consumer_twin_id,
+            Error::<T>::TwinNotAuthorizedToSetMetadata,
+        );
+
+        // Only allow to modify metadata if contract still not approved by both parties
+        ensure!(
+            !matches!(
+                service_contract.state,
+                types::ServiceContractState::ApprovedByBoth
+            ),
+            Error::<T>::ServiceContractModificationNotAllowed,
+        );
+
+        // TODO create metadata input type to control size
+        service_contract.metadata = BoundedVec::try_from(metadata)
+            .map_err(|_| Error::<T>::ServiceContractMetadataTooLong)?;
+
+        // If base_fee is set and non-zero (mandatory)
+        if service_contract.base_fee != 0 {
+            service_contract.state = types::ServiceContractState::AgreementReady;
```

Comment:
I can still get the contract state to `AgreementReady` if I set the `base_fee` but skip the metadata? (see below)

Reply:
No, both `base_fee` and `metadata` need to be filled to get `AgreementReady` state
according specs https://github.com/threefoldtech/tfchain/issues/445#issuecomment-1280725022